### PR TITLE
Generate docs and schemas from common schema definition format

### LINKED ISSUE #282: Documentation and schemas should be generated from a common source
### Description
Today the schemas and the Markdown documentation files are maintained separately by hand, thus spreading out information about events. We should collect that information in a single place and generate both schemas and documentation files.

I don't have a particular file format in mind, but one _could_ imagine a YAML format that's more or less a superset of the JSON schema specification with additional keys for documentation and other metadata used when generating documentation but stripped when generating a schema.

### Motivation
* The schemas can't express all rules in the spec, like what event types are valid targets for particular link types. Right now the schemas don't even express which link types are valid for an event type, but at least that's fixable (see #148).
* Changing parts of the documentation shared between all events requires modifying and reviewing 23 files.
* The lack of machine-readable documentation of the event members makes it hard to generate SDKs with documentation (Javadoc, docstrings, godoc, ...).
* In #281 there's an excellent suggestion to visualize the relationships between event types but that would currently require manual labor.
* When e.g. considering protocol changes it's often useful to understand how events can relate, like "what kind of events can link to CD?". Right now the best way of answering such a question is probably to grep the Markdown files and that's not a fantastic method.

### Exemplification
See above.

### Benefits
The current representation of event schemas and documentation is an obstacle for protocol maintainers and others who want to process the information with a program. Making this change would greatly improve the situation.

### Possible Drawbacks
Complexity? Hand-written files are labor intensive but they're at least simple to grasp.

### PR DESCRIPTION
**Pull request by magnusbaeck**
### Applicable Issues
Fixes #282

### Description of the Change
A new schema definition format that's a superset of the normal JSON schema format is introduced and all existing schemas and documentation files are merged into files in this new file format, stored in the "definitions" directory.

At the same time, the definitions of the meta member and event links are extracted to their own definition files and referenced from the event files. This both allows reuse but also makes it trivial for scripts like SDK generators to understand the structure better. Since both meta and links have evolved over time we've reconstructed different versions of those subschemas.

The schema files aren't affected at all by these changes, i.e. running generate_schemas.py on a definition file will produce a byte-identical schema file. This isn't quite true for the .md documentation; some minor changes are introduced:

- Whitespace and other things that doesn't affect the rendering.
- Some link types are reordered so they're in alphabetical order.
- Previously some key/value pairs listed for a member ("Type",      "Required", "Description", etc) were present but empty. Now only      non-empty values cause the key to get included at all.
- Members below meta.security.integrityProtection and a few other      members for other events are now listed in a slightly different       order (namely the one from the JSON schema).

To make sure the definition files don't diverge from the generated files we have an additional job in the GitHub Actions workflow that's run for all pushes and PRs. It runs the generation scripts and verifies that we won't get any dirty files in the workspace and no untracked files. See https://github.com/magnusbaeck/eiffel/actions/runs/3007678382 and https://github.com/magnusbaeck/eiffel/actions/runs/3007462169 for examples of what violations look like.

### Alternate Designs
A few other options were discussed in the associated issue.

### Benefits
The current representation of event schemas and documentation is an obstacle for protocol maintainers and others who want to process the information with a program. Making this change would greatly improve the situation.

### Possible Drawbacks
None, I think.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I    have the right to submit it under the open source license    indicated in the file; or

(b) The contribution is based upon previous work that, to the best    of my knowledge, is covered under an appropriate open source    license and I have the right under that license to submit that    work with modifications, whether created in whole or in part    by me, under the same open source license (unless I am    permitted to submit under a different license), as indicated    in the file; or

(c) The contribution was provided directly to me by some other    person who certified (a), (b) or (c) and I have not modified    it.

(d) I understand and agree that this project and the contribution    are public and that a record of the contribution (including all    personal information I submit with it, including my sign-off) is    maintained indefinitely and may be redistributed consistent with    this project or the open source license(s) involved.

Signed-off-by: Magnus Bäck \<magnus.back@axis.com>

### CODE REVIEW

**Comment by e-backmark-ericsson** [comment]
Sorry for raising this concern now, I'm either forgetting some good argumentation we've had or I've understood the consequences of this change too late.

With this solution we still need to create _new files_ for a new version of each event. I do believe the reason for having individual json schemas for all historical event version stored in the repo was to make it easy to reference those schemas for all historical versions through one single branch of the repo. Now that we move from having the json schemas as the source of truth, and instead generate them for each new release (or each new version of the events?), do we still need this or could we let the yaml files be the same and change them for each new version instead?

**Comment by magnusbaeck** [comment]
It seems like orthogonal questions:

- Should the schema files and documentation be hand-edited or generated from a single source?
- Should new versions result in new files or changes in existing files?

Schema files need to be regenerated and committed along with schema the definition file so this PR doesn't change the structure of the repo and how it's consumed.

**Comment by e-backmark-ericsson** [comment]
Well the first question is what this PR and its related issue is all about, so that is of no concern I believe.
The second question has a follow up question hidden in it - how do we make sure the generated json files are aligned with the commited yaml files? If we don't at the same time introduce a new structure, for example by _not_ checking in those json schemas but instead generate them at merge of a PR, we instead need to add a sanity test to verify that the manually generated json schemas match the commited yaml files. It won't be sustainable to review such differences manually for each change to the protocol.
And I believe that the individual json schemas per version could still be generated even if we don't create new yaml files for each version, by reading old commits from the repo, or?

**Comment by e-backmark-ericsson** [comment]
As long as the validation of the generated json files match the yaml files is part of the Github action checking any protocol PRs, we don't need to introduce any more improvements through this PR.
In the long run we might want to generate an external documentation site for our events, and then we will have a lot of new requirements to deal with there.

**Comment by z-sztrom** [comment]
Works fine for me. Both schema and documentation generated properly.

**Comment by z-sztrom** [comment]
I cannot review the changes. No enough permissions?

**Comment by magnusbaeck** [comment]
Weird, you should be able to review the PR (but since you're not a maintainer any approval you make won't count towards any minimum number of approvals).

**Review by z-sztrom — COMMENTED** [review]
Nice work! Generated schemas equal to the schemas I modified manually some time ago.

**Comment by magnusbaeck** [comment]
One thing I don't think we've discussed is what to do about changes to the YAML files that don't affect the generated schemas, like documentation changes and changes in the links. For example, #287 adds a new link type to a few events types which in that PR amounts to a mere documentation change, but when merging that change into this PR I can only assume that we need to issue new event versions (with a minor version bump)? Those schemas would currently be identical to the preceding versions.

**Comment by magnusbaeck** [comment]
No, wait. When merging the PRECURSOR branch (or master, after that PR is merged) into this PR branch the current events should be regarded as settled law and we obviously shouldn't issue new event versions after the fact, but what should we do _the next time_ someone adds a link type or makes a similar non-trivial change to a schema definition file that doesn't affect the schema?

**Comment by magnusbaeck** [comment]
This PR is now up to date with @e-backmark-ericsson's recently merged PRECURSOR change so it's ready for a review. Let me know if there would be any point in a "group review" where we go through things together.

**Comment by e-backmark-ericsson** [comment]
> No, wait. When merging the PRECURSOR branch (or master, after that PR is merged) into this PR branch the current events should be regarded as settled law and we obviously shouldn't issue new event versions after the fact, but what should we do _the next time_ someone adds a link type or makes a similar non-trivial change to a schema definition file that doesn't affect the schema?

I'd say that the minor version of the event should be stepped if there is a link-only update to it, even though we (currently) don't check such things in the schemas. The reason for not having that as part of the schemas is that when the schemas were originally created there were no known means to validate that a list of dictionaries actually had certain keys in their dictionaries. I believe that more recent json schema version can handle that, and as soon as we add such restrictions to the schemas they will need to be updated also for link-only updates in the events.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelIssueDefinedEvent.md:2]: As commented in other places, this doesn't really match the original. Should we instead provide the copyright header as a value to a key with underscore in the yaml files?

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelIssueDefinedEvent.md:19]:
```suggestion
The EiffelIssueDefinedEvent declares that an issue has been created in some external issue management software. ID is semantically similar to [EiffelFlowContextDefinedEvent](../eiffel-vocabulary/EiffelFlowContextDefinedEvent.md) and [EiffelEnvironmentDefinedEvent](../eiffel-vocabulary/EiffelEnvironmentDefinedEvent.md).
```

File: definitions/EiffelIssueDefinedEvent/1.0.0.yml
Comment (e-backmark-ericsson) [inline definitions/EiffelIssueDefinedEvent/1.0.0.yml:4]:
```suggestion
  been created in some external issue management software. ID is semantically
```

File: definitions/EiffelIssueDefinedEvent/2.0.0.yml
Comment (e-backmark-ericsson) [inline definitions/EiffelIssueDefinedEvent/2.0.0.yml:4]:
```suggestion
  been created in some external issue management software. ID is semantically
```

File: definitions/EiffelIssueDefinedEvent/3.0.0.yml
Comment (e-backmark-ericsson) [inline definitions/EiffelIssueDefinedEvent/3.0.0.yml:4]:
```suggestion
  been created in some external issue management software. ID is semantically
```

File: definitions/EiffelIssueDefinedEvent/3.1.0.yml
Comment (e-backmark-ericsson) [inline definitions/EiffelIssueDefinedEvent/3.1.0.yml:4]:
```suggestion
  been created in some external issue management software. ID is semantically
```

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md:19]: There are multiple whitespace issues in this file, emerging from the yaml files. Please update those.

**Review by e-backmark-ericsson — CHANGES_REQUESTED** [review]
Overall I think this PR looks very good. Job well done! I have a few requests for changes though. Most of them are easily fixed.

Regarding the comments on the copyright headers - I believe the correct way to handle those is to actually move the copyright headers out of the generated md files and instead provide them as comments in the yml files. There should really bee no need to make them part of the generated md files by adding them as a key with underscore to the yml file structure. The actual copyright should hold for the file that is manually edited (the yml file) and not for the generated files.

File: eiffel-syntax-and-usage/event-schemas.md
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/event-schemas.md:35]: No validation takes place. I started working on a JSON schema for the YAML files but found it surprisingly hard (unless we want to duplicate the JSON schema schema and add some additional valid keys). I was hoping we could add such a schema as a follow-up.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (magnusbaeck) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md]: Fixed in 888fdaa893e88ed28b6374edf2a6aa24539ccfec.

File: definitions/EiffelIssueDefinedEvent/1.0.0.yml
Comment (magnusbaeck) [reply definitions/EiffelIssueDefinedEvent/1.0.0.yml]: Fixed in 888fdaa893e88ed28b6374edf2a6aa24539ccfec.

File: definitions/EiffelIssueDefinedEvent/2.0.0.yml
Comment (magnusbaeck) [reply definitions/EiffelIssueDefinedEvent/2.0.0.yml]: Fixed in 888fdaa893e88ed28b6374edf2a6aa24539ccfec.

File: eiffel-syntax-and-usage/event-schemas.md
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/event-schemas.md]: Fixed in 876abda2c183aedad95e25da8652750e9e4ca48f.

File: eiffel-syntax-and-usage/event-schemas.md
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/event-schemas.md:81]: No reason not to do it, really. I guess it just didn't occur to me. I'll try to push a commit with that tomorrow.

**Comment by magnusbaeck** [comment]
> I'd say that the minor version of the event should be stepped if there is a link-only update to it, even though we (currently) don't check such things in the schemas.

Yeah, I agree. I'll update eiffel-syntax-and-usage/versioning.md to document this.

> Regarding the comments on the copyright headers - I believe the correct way to handle those is to actually move the copyright headers out of the generated md files and instead provide them as comments in the yml files. There should really bee no need to make them part of the generated md files by adding them as a key with underscore to the yml file structure. The actual copyright should hold for the file that is manually edited (the yml file) and not for the generated files.

Yep. There was no good reason for omitting the copyright notice from the YAML files. We _could_ generate a file header for the .md files that states the origin of the file, something like this:

```
<!--
This file was generated from ../definitions/EiffelCompositionDefinedEvent/3.2.0.yml.
See that file for a copyright notice.
-->
```

File: eiffel-syntax-and-usage/event-schemas.md
Comment (e-backmark-ericsson) [reply eiffel-syntax-and-usage/event-schemas.md:35]: Sure. Let's bring that in later if we see a critical need for it.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelIssueDefinedEvent.md:69]:
```suggestion
__Description:__ Identifies a cause of the event occurring. SHOULD not be used in conjunction with __CONTEXT__: individual events providing __CAUSE__ within a larger context gives rise to ambiguity. It is instead recommended to let the root event of the context declare __CAUSE__.
```

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelIssueDefinedEvent.md:75]:
```suggestion
__Description:__ Identifies the activity or test suite of which this event constitutes a part.
```

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelIssueDefinedEvent.md:81]:
```suggestion
__Description:__ Identifies the flow context of the event: which is the continuous integration and delivery flow in which this occurred – e.g. which product, project, track or version this is applicable to.
```

**Review by e-backmark-ericsson — CHANGES_REQUESTED** [review]
Some more whitespaces missing

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (e-backmark-ericsson) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md]: Some more whitespace errors spotted. Commented separately.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (magnusbaeck) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md]: Fixed in 837dc1f55476b3ca0ba2d9e8c235f39c4cd7ecf3.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (magnusbaeck) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md]: Fixed in 837dc1f55476b3ca0ba2d9e8c235f39c4cd7ecf3.

File: eiffel-vocabulary/EiffelIssueDefinedEvent.md
Comment (magnusbaeck) [reply eiffel-vocabulary/EiffelIssueDefinedEvent.md]: Fixed in 837dc1f55476b3ca0ba2d9e8c235f39c4cd7ecf3.

File: event_docs.md.j2
Comment (magnusbaeck) [reply event_docs.md.j2]: Fixed in 28abb25871ead9c6da4e36b9bacef615be29134d.

**Comment by magnusbaeck** [comment]
> Yeah, I agree. I'll update eiffel-syntax-and-usage/versioning.md to document this.

Done in 0fef4211126e7e628bc1fc6f50359b5a5a156fd3.

> Yep. There was no good reason for omitting the copyright notice from the YAML files. We _could_ generate a file header for the .md files that states the origin of the file, something like this:

Done in 28abb25871ead9c6da4e36b9bacef615be29134d.

File: eiffel-syntax-and-usage/event-schemas.md
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/event-schemas.md:81]: Done in b6b4bc50b470532061ee71359cba20213fbd196e.

**Review by e-backmark-ericsson — APPROVED** [review]
LGTM

**Review by t-persson — APPROVED** [review]

**Comment by magnusbaeck** [comment]
@z-sztrom, this PR has been merged now so feel free to rebase and upload your schema URL PR whenever you have time.

**Pull request merged**
